src/components/RelicModal.jsx

```diff
@@ -303,7 +478,344 @@ export default function RelicModal(props) {
 
   const plusThree = () => {
-    relicForm.setFieldValue("enhance", Math.min(relicForm.getFieldValue("enhance") + 3, 15));
+    relicForm.setFieldValue('enhance', Math.min(relicForm.getFieldValue('enhance') + 3, 15))
+  }
+
+  const [wearerID, setWearerID] = useState('None')
+
+  const [hide0, setHide0] = useState('inline')
+
+  const [hide1, setHide1] = useState('inline')
+
+  const [hide2, setHide2] = useState('inline')
+
+  const [hide3, setHide3] = useState('inline')
+
+  const [button0low, setButton0low] = useState(undefined)
+
+  const [button0mid, setButton0mid] = useState(undefined)
+
+  const [button0high, setButton0high] = useState(undefined)
+
+  const [button1low, setButton1low] = useState(undefined)
+
+  const [button1mid, setButton1mid] = useState(undefined)
+
+  const [button1high, setButton1high] = useState(undefined)
+
+  const [button2low, setButton2low] = useState(undefined)
+
+  const [button2mid, setButton2mid] = useState(undefined)
+
+  const [button2high, setButton2high] = useState(undefined)
+
+  const [button3low, setButton3low] = useState(undefined)
+
+  const [button3mid, setButton3mid] = useState(undefined)
+
+  const [button3high, setButton3high] = useState(undefined)
+
+  const [trueStats, setTrueStats] = useState({
+    substat0: 0,
+    substat1: 0,
+    substat2: 0,
+    substat3: 0,
+  })
+
+  const [precision0, setPrecision0] = useState(0)
+
+  const [precision1, setPrecision1] = useState(0)
+
+  const [precision2, setPrecision2] = useState(0)
+
+  const [precision3, setPrecision3] = useState(0)
+
+  const upgrade0low = () => {
+    const stat = relicForm.getFieldValue('substatType0')
+    const grade = relicForm.getFieldValue('grade')
+    const increment = SubStatValues[stat][grade].low
+    const stats = trueStats
+    if (stat == Constants.Stats.SPD) {
+      stats.substat0 += 2
+      setButton0low(Math.floor(stats.substat0) + 2)
+      setButton0high(Math.floor(stats.substat0) + 3)
+      relicForm.setFieldValue('substatValue0', Math.floor(stats.substat0))
+    } else if (Utils.isFlat(stat)) {
+      stats.substat0 += increment
+      setButton0low(Math.floor(stats.substat0 + SubStatValues[stat][grade].low))
+      setButton0mid(Math.floor(stats.substat0 + SubStatValues[stat][grade].mid))
+      setButton0high(Math.floor(stats.substat0 + SubStatValues[stat][grade].high))
+      relicForm.setFieldValue('substatValue0', Math.floor(stats.substat0))
+    } else {
+      stats.substat0 += increment
+      setButton0low(parseFloat(Math.floor(10 * (stats.substat0 + SubStatValues[stat][grade].low)) / 10).toFixed(1))
+      setButton0mid(parseFloat(Math.floor(10 * (stats.substat0 + SubStatValues[stat][grade].mid)) / 10).toFixed(1))
+      setButton0high(parseFloat(Math.floor(10 * (stats.substat0 + SubStatValues[stat][grade].high)) / 10).toFixed(1))
+      relicForm.setFieldValue('substatValue0', parseFloat(Math.floor(10 * stats.substat0) / 10).toFixed(1))
+    }
+    plusThree()
+  }
```

Comment:
Way too much code duplication, we have 12 blocks of code that do almost exactly the same thing here. These identical blocks should probably be functions & reused